### PR TITLE
Stencil Params: Add Stencil Compare and Write Mask Material Parameters

### LINKED ISSUE #17419: Material has no control over stencilBuffer.setMask()
##### Description of the problem

After the recent changes that offer control over the stencil buffer state through the material interface in #15611 , one major aspect of the stencil state setup was overlooked: the stencil write mask.

This can be set directly through the stencil buffer like: 
``` javascript
renderer.state.buffers.stencil.setMask(MASK);
```

But, as we can see [here](https://github.com/mrdoob/three.js/blob/master/src/renderers/webgl/WebGLState.js#L684), it's not exposed nor used through the material's stencil parameters:
``` javascript
var stencilWrite = material.stencilWrite;
stencilBuffer.setTest( stencilWrite );
if ( stencilWrite ) {
	stencilBuffer.setFunc( material.stencilFunc, material.stencilRef, material.stencilMask );
	stencilBuffer.setOp( material.stencilFail, material.stencilZFail, material.stencilZPass );
}
```

Material's ``material.stencilMask`` should be split in ``material.stencilWriteMask`` (which would be use in conjunction with ``renderer.state.buffers.stencil.setMask()``) and ``material.stencilFuncMask`` to properly expose all the stencil state functionality through the material layer.

Some nice links on the topic:
* [stencilMask](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/stencilMask)
* [stencilFunc](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/stencilFunc)

##### Three.js version

- [X] Dev
- [X] r108
- [ ] ...

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### PR DESCRIPTION
**Pull request by gkjohnson**
Fix #17419

Changes `stencilMask` to `stencilFuncMask` and adds `stencilWriteMask` (corresponding to the [stencilMask](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/stencilMask) function).

For the moment the setter and getter for `stencilMask` will log a warning they are used.

### CODE REVIEW

File: src/materials/Material.js [hunk 1]
```diff
@@ -41,14 +41,33 @@ function Material() {
...
+
+		get: function () {
+
+			console.warn( 'Material.stencilMask has been removed. Use Material.stencilFuncMask instead.' );
+			return this.stencilFuncMask;
+
+		}
+
+	} );
+
```
Comment (Mugen87) [inline src/materials/Material.js]: Stuff like that normally goes to `three.legacy.js`.

File: src/materials/Material.js [hunk 2]
```diff
@@ -362,9 +381,10 @@ Material.prototype = Object.assign( Object.create( EventDispatcher.prototype ),
 		this.depthWrite = source.depthWrite;
 
 		this.stencilWrite = source.stencilWrite;
+		this.stencilWriteMask = source.stencilWriteMask;
 		this.stencilFunc = source.stencilFunc;
 		this.stencilRef = source.stencilRef;
-		this.stencilMask = source.stencilMask;
+		this.stencilFuncMask = source.stencilFuncMask;
```
Comment (Mugen87) [inline src/materials/Material.js]: It might also be a good idea to enhance `MaterialLoader` so these properties can be actually deserialized.

**Comment by Mugen87** [comment]
It seems you also have to change `Material.toJSON()`, see

https://github.com/mrdoob/three.js/pull/17449/files#diff-f5eb8fcc970231604c15522381d9f59eR282

**Comment by gkjohnson** [comment]
@Mugen87 Thanks I've updated `toJSON`, moved the property warning to `Three.Legacy.js` and updated `MaterialLoader` to deserialize stencil properties!

**Review by Mugen87 — APPROVED** [review]

**Comment by mrdoob** [comment]
Thanks!

**Pull request merged**